### PR TITLE
Adds a tool to generate ground truth data.

### LINKED ISSUE #310: Semi-automatic generation of ground truth data
Ideally ground truth data is generated just using the serialized output in the `SparsePoseGraph` proto. 

It is difficult to visually assess the SLAM quality over short periods of time, i.e. to decide whether a map is somehow warped. It is easier to assess whether good loop closures were found. Using a single trajectory, one approach might be the following:

* look at all loop closure constraints, i.e. `INTER_SUBMAP`
* suitable constraints should be separated by a long covered distance, i.e. not by waiting a long time
* suitable constraints should be almost satisfied in the end, i.e. not outliers
* for each such constraints
  1. pick a point in time during the submap was generated, e.g. choose the `INTRA_SUBMAP` scan in the middle of the submap
  1. compute the relative transform between this scan of the submap and the scan of the constraint
  1. store both times and the relative transform

### PR DESCRIPTION
**Pull request by wohe**
Fixes #310. Not yet tested.

### CODE REVIEW

File: cartographer/ground_truth/autogenerate_ground_truth_main.cc [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+#include "cartographer/mapping/proto/sparse_pose_graph.pb.h"
+#include "cartographer/transform/transform.h"
+#include "gflags/gflags.h"
+#include "glog/logging.h"
+
+DEFINE_string(
+    pose_graph_filename, "",
+    "File with the pose graph proto from which to generate ground truth data.");
+DEFINE_string(output_filename, "", "File to write the ground truth proto to.");
+DEFINE_double(min_covered_distance, 100.,
```
Comment (SirVer) [inline cartographer/ground_truth/autogenerate_ground_truth_main.cc]: unit?

File: cartographer/ground_truth/autogenerate_ground_truth_main.cc [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+              "candidate for autogenerated ground truth.");
+DEFINE_double(outlier_threshold, 0.15,
+              "Distance in meters beyond which constraints are considered "
+              "outliers.");
+
+namespace cartographer {
+namespace ground_truth {
+namespace {
+
+std::vector<double> ComputeCoveredDistance(
```
Comment (SirVer) [inline cartographer/ground_truth/autogenerate_ground_truth_main.cc]: ComputeCummulativeCoveredDistance?

File: cartographer/ground_truth/autogenerate_ground_truth_main.cc [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    const auto last_pose = transform::ToRigid3(trajectory.node(i - 1).pose());
+    const auto this_pose = transform::ToRigid3(trajectory.node(i).pose());
+    covered_distance.push_back(
+        covered_distance.back() +
+        (last_pose.inverse() * this_pose).translation().norm());
+  }
+  return covered_distance;
+}
+
+// We pick the representative in the middle of the submap.
```
Comment (SirVer) [inline cartographer/ground_truth/autogenerate_ground_truth_main.cc]: What is the representative? Representative what?

File: cartographer/ground_truth/autogenerate_ground_truth_main.cc [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    const int next_submap_index = static_cast<int>(submap_to_scan_index.size());
+    const int submap_index = constraint.submap_id().submap_index();
+    if (submap_index <= next_submap_index) {
+      continue;
+    }
+
+    CHECK_EQ(submap_index, next_submap_index + 1);
+    submap_to_scan_index.push_back(constraint.scan_id().scan_index());
+  }
+
```
Comment (SirVer) [inline cartographer/ground_truth/autogenerate_ground_truth_main.cc]: remove empty line

File: cartographer/ground_truth/autogenerate_ground_truth_main.cc [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+  int num_outliers = 0;
+  proto::GroundTruth ground_truth;
+  for (const auto& constraint : pose_graph.constraint()) {
+    // We're only interested in loop closure constraints.
+    if (constraint.tag() ==
+        mapping::proto::SparsePoseGraph::Constraint::INTRA_SUBMAP) {
+      continue;
+    }
+
+    // A representative node for the submap must have been chosen, i.e. not be
```
Comment (SirVer) [inline cartographer/ground_truth/autogenerate_ground_truth_main.cc]: I do not understand this comment.

File: cartographer/ground_truth/autogenerate_ground_truth_main.cc [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+int main(int argc, char** argv) {
+  google::InitGoogleLogging(argv[0]);
+  FLAGS_logtostderr = true;
+  google::SetUsageMessage(
+      "\n\n"
+      "This program semi-automatically generates ground truth data from a\n"
+      "pose graph proto.\n"
+      "\n"
+      "The input should contain a single trajectory and should have been\n"
+      "manually assessed to be correctly loop closed. Small local distortions\n"
```
Comment (SirVer) [inline cartographer/ground_truth/autogenerate_ground_truth_main.cc]: What do you mean by Small local distortions? What is small?

File: cartographer/ground_truth/autogenerate_ground_truth_main.cc (hunk 2 above)
Comment (wohe) [inline cartographer/ground_truth/autogenerate_ground_truth_main.cc]: Seems redundant to me. `ComputeCoveredDistanceFromStart()` maybe? I'd favor conciseness, but let me know if you would like me to add `FromStart`.

**Pull request merged**